8355003: Implement JEP 515: Ahead-of-Time Method Profiling

veresov opened the pull request:
Improve warm-up time by making profile data from a previous run of an application instantly available, when the HotSpot Java Virtual Machine starts. Specifically, enhance the [AOT cache](https://openjdk.org/jeps/483) to store method execution profiles from training runs, reducing profiling delays in subsequent production runs.

More details in the JEP: https://bugs.openjdk.org/browse/JDK-8325147

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Change requires CSR request [JDK-8356012](https://bugs.openjdk.org/browse/JDK-8356012) to be approved
- [x] Commit message must refer to an issue

### Issues
 * [JDK-8355003](https://bugs.openjdk.org/browse/JDK-8355003): Implement JEP 515: Ahead-of-Time Method Profiling (**Sub-task** - P4)
 * [JDK-8356012](https://bugs.openjdk.org/browse/JDK-8356012): Implement JEP 515: Ahead-of-Time Method Profiling (**CSR**)


### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)
 * [Magnus Ihse Bursie](https://openjdk.org/census#ihse) (@magicus - **Reviewer**) Review applies to [7fb7ae62](https://git.openjdk.org/jdk/pull/24886/files/7fb7ae62784cc7f43eecfdb00390e6a48feede27)
 * [Chris Plummer](https://openjdk.org/census#cjplummer) (@plummercj - **Reviewer**) Review applies to [ae332887](https://git.openjdk.org/jdk/pull/24886/files/ae332887801d154959fb1f77fc14a7f3c4ec4620)
 * [Ioi Lam](https://openjdk.org/census#iklam) (@iklam - **Reviewer**) Review applies to [d3d51b00](https://git.openjdk.org/jdk/pull/24886/files/d3d51b0007ce2def1556fa114e3af868160980c3)

### Contributors
 * John R Rose `<jrose@openjdk.org>`
 * Vladimir Ivanov `<vlivanov@openjdk.org>`
 * Ioi Lam `<iklam@openjdk.org>`
 * Vladimir Kozlov `<kvn@openjdk.org>`
 * Aleksey Shipilev `<shade@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/24886/head:pull/24886` \
`$ git checkout pull/24886`

Update a local copy of the PR: \
`$ git checkout pull/24886` \
`$ git pull https://git.openjdk.org/jdk.git pull/24886/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 24886`

View PR using the GUI difftool: \
`$ git pr show -t 24886`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/24886.diff">https://git.openjdk.org/jdk/pull/24886.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/24886#issuecomment-2831378242)
</details>

veresov commented:
/label hotspot-compiler
/label hotspot-runtime

veresov commented:
/contributor jrose
/contributor vlivanov
/contributor iklam
/contributor kvn
/contributor shade

veresov commented:
/contributor add jrose
/contributor add vlivanov
/contributor add iklam
/contributor add kvn
/contributor add shade

vnkozlov commented on src/hotspot/share/ci/ciMethodData.cpp:
```diff
@@ -68,7 +68,11 @@ class PrepareExtraDataClosure : public CleanExtraDataClosure {
 
   bool is_live(Method* m) {
-    if (!m->method_holder()->is_loader_alive()) {
+    Klass* holder = m->method_holder();
```
Changes in this file seems not related and can be pushed/tested separately. If they are related - there should be condition for additional checks.

veresov replied:
You mean you want these checks to be done only if `TrainingData::have_data() == true` ?

vnkozlov replied:
Yes, if it is related. Otherwise you may change default behavior when Leyden code is not used.

veresov replied:
I simplified it and factored out the new semantics.

vnkozlov commented on src/hotspot/share/ci/ciMethod.cpp:
```diff
@@ -1142,6 +1144,28 @@ int ciMethod::code_size_for_inlining() {
 // Also some instructions inside the code are excluded from inline
 // heuristic (e.g. post call nop instructions; see InlineSkippedInstructionsCounter)
 int ciMethod::inline_instructions_size() {
+  if (_inline_instructions_size == -1) {
```
Why repeat this condition and not put new code under existing one?

veresov replied:
The caching logic sets the `_inline_instructions_size` if the value is found in the archive. The normal logic doesn't need to run if this happens. 
Something like:
```
  if (_value == -1) {
       _value = get_from_cache();
  }
  if (_value == -1) { // didn't get it from the cache
    _value = compute_value();
  }
```
So basically to delineate the caching logic from the normal path. Does it make sense?

vnkozlov commented on src/hotspot/share/cds/dumpAllocStats.hpp:
```diff
@@ -146,6 +148,11 @@ class DumpAllocStats : public StackObj {
     _num_method_cp_entries_reverted += reverted ? 1 : 0;
   }
 
+  void record_dynamic_proxy_class() {
```
This is not called. This code seems not related.

veresov replied:
True. @iklam, this came with a change you wanted me to take. Ok to cut this out?

iklam replied:
Yes, this part is not needed. AOT dynamic proxy classes are only supported in the Leyden repo.

vnkozlov commented on src/hotspot/share/oops/trainingData.hpp:
```diff
@@ -0,0 +1,835 @@
+  bool saw_level(CompLevel l) const { return (_level_mask & level_mask(l)) != 0; }
+  int highest_level()         const { return highest_level(_level_mask); }
+  int highest_top_level()     const { return _highest_top_level; }
+  MethodData* final_profile() const { return _final_profile; }
```
`never_inlined()`, `highest_level()` are not used.

vnkozlov commented on src/hotspot/share/ci/ciObjectFactory.hpp:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 1999, 2025, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 1999, 2020, Oracle and/or its affiliates. All rights reserved.
```
Wrong year

vnkozlov commented on src/hotspot/share/ci/ciInstanceKlass.hpp:
```diff
@@ -44,6 +44,7 @@ class ciInstanceKlass : public ciKlass {
   friend class ciMethod;
   friend class ciField;
   friend class ciReplay;
+  friend class CompileTrainingData;
```
Not referenced here

veresov replied:
It allows `CompileTrainingData` to peek into the `ciInstanceKlass` internals. We need the klass ptr specially.

vnkozlov replied:
I missed that it is "friend" declaration.

vnkozlov commented on src/hotspot/share/cds/archiveBuilder.cpp:
```diff
@@ -739,6 +767,10 @@ void ArchiveBuilder::relocate_metaspaceobj_embedded_pointers() {
   log_info(cds)("Relocating embedded pointers in core regions ... ");
   relocate_embedded_pointers(&_rw_src_objs);
   relocate_embedded_pointers(&_ro_src_objs);
+  log_info(cds)("Relocating %zu pointers, %zu tagged, %zu nulled",
```
`log_info(aot)` if it is Leyden related.

veresov replied:
This more like a generic cds message. I'll leave this one as is.

vnkozlov commented on src/hotspot/share/oops/trainingData.cpp:
```diff
@@ -0,0 +1,802 @@
+/*
+ * Copyright (c) 2023, 2025, Oracle and/or its affiliates. All rights reserved.
```
Use one year

vnkozlov commented on src/hotspot/share/oops/trainingData.cpp:
```diff
@@ -0,0 +1,802 @@
+volatile bool TrainingData::TrainingDataLocker::_snapshot = false;
+
+MethodTrainingData::MethodTrainingData() {
+  assert(CDSConfig::is_dumping_static_archive() || UseSharedSpaces, "only for CDS");
```
Consider adding and using `CDSConfig::is_dumping_training_data() ` or something.

vnkozlov commented on src/hotspot/share/oops/trainingData.cpp:
```diff
@@ -0,0 +1,802 @@
+}
+
+static void verify_archived_entry(TrainingData* td, const TrainingData::Key* k) {
+  guarantee(TrainingData::Key::can_compute_cds_hash(k), "");
```
Should we gracefully disable using TD instead of crashing VM?

veresov replied:
But this is a verification code. That seems to be the usual strategy, is it not?

vnkozlov replied:
I thought if we can not use AOT cache we issue warning and continue execution without it.
Unless you are checking for damaged AOT cache which may affect execution without it.

@iklam  what do you think?

veresov replied:
But this runs only with `-XX:+VerifyTrainingData`. It's a testing mode.

vnkozlov replied:
I missed that. Okay then.

vnkozlov commented on src/hotspot/share/oops/trainingData.cpp:
```diff
@@ -0,0 +1,802 @@
+    bool is_excluded = !holder()->is_loaded() || SystemDictionaryShared::check_for_exclusion(holder(), nullptr);
+    if (is_excluded) {
+      ResourceMark rm;
+      log_debug(cds)("Cleanup KTD %s", name()->as_klass_external_name());
```
`log_debug(aot, training)`

vnkozlov reviewed (changes requested):
In general it is okay.

Please, use UL with `(aot, training)` tags for your code.

I noticed (and added comment) that you use `guarantee()` which crashes VM when you call new `verify` methods. Can you disable TD instead and continue execution?

veresov commented on test/hotspot/jtreg/runtime/cds/appcds/aotProfile/AOTProfileFlags.java:
```diff
@@ -0,0 +1,61 @@
+ * @summary Sanity test of combinations of the diagnostic flags [+-]AOTRecordTraining and [+-]AOTReplayTraining
+ * @requires vm.cds
+ * @comment work around JDK-8345635
+ * @requires !vm.jvmci.enabled
```
Could you please explain why? @iklam, what do you think?

iklam replied:
I think it's OK to test without these two additional flags. This will make sure that the two diagnostic flags don't have any bad side effect even if AOT class linking is disabled (due to flags like -XX:+UseZGC).

vnkozlov replied:
I see them in `aotClassLinking/AOTClassLinkingVMOptions.java` test.
@veresov please, add link to your mach5 testing to JBS. And add to description what tests you already ran.

iklam replied:
OK, let's add those two `@requires` to be consistent with other tests.

veresov commented on src/hotspot/share/runtime/init.cpp:
```diff
@@ -185,6 +186,10 @@ jint init_globals2() {
   }
 #endif
 
+  if (TrainingData::have_data() || TrainingData::need_data()) {
```
Because it's only needed if we're recording or replaying. I'll add a comment.

veresov commented on src/hotspot/share/oops/methodData.cpp:
```diff
@@ -1364,7 +1431,7 @@ void MethodData::init() {
 
 bool MethodData::is_mature() const {
-  return CompilationPolicy::is_mature(_method);
+  return CompilationPolicy::is_mature((MethodData*)this);
```
To remove constness. I'll make it a const_cast.

veresov commented on src/hotspot/share/oops/methodData.cpp:
```diff
@@ -1223,10 +1286,14 @@ void MethodData::post_initialize(BytecodeStream* stream) {
+}
+
+MethodData::MethodData() {
+  assert(CDSConfig::is_dumping_static_archive() || UseSharedSpaces, "only for CDS");
```
I think those are used for CDS serialization/deserialization, right, @iklam?

iklam replied:
This constructor is used by cppVtables.cpp to calculate the size of the vtables for MethodData, and also for finding the address of the vtable of  MethodData. All types in `CPP_VTABLE_TYPES_DO` must have such an empty constructor. E.g., `InstanceKlass::InstanceKlass()`.

We have not been very consistent with comments around these constructors, but I think we can do this:

```
#if INCLUDE_CDS
MethodData::MethodData() {
  // Used by cppVtables.cpp only
  assert(CDSConfig::is_dumping_static_archive() || UseSharedSpaces, "only for CDS");
}
#endif
```

This method is called even if we are not dumping training data. The vtables of all types in `CPP_VTABLE_TYPES_DO` are unconditionally computed.

veresov replied:
@vnkozlov Are you ok with Ioi's code change proposal?

vnkozlov replied:
Yes, I am ok. I assume you will change accordingly all other similar empty constructors in your code: `MethodCounters()` and `*TrainingData()`.

veresov replied:
Yes, will do.

erikj79 commented:
Build changes look trivially good.

vnkozlov reviewed (commented):
Looks better.
There are still places where UL is used specifically for TD processing. Consider using `(aot, training)` there instead of `(cds)`.

veresov commented:
> Looks better. There are still places where UL is used specifically for TD processing. Consider using `(aot, training)` there instead of `(cds)`.

Right. I haven't addressed these review comments yet.

magicus reviewed (approved):
Build changes are trivially fine.

veresov commented:
Vladimir (@vnkozlov), I did the changes that you requested. You please do another pass? Thanks!

vnkozlov reviewed (approved):
Looks good.

plummercj commented on src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/runtime/Threads.java:
```diff
@@ -151,6 +151,7 @@ private static synchronized void initialize(TypeDataBase db) {
 
         if (!VM.getVM().isCore()) {
             virtualConstructor.addMapping("CompilerThread", CompilerThread.class);
+            virtualConstructor.addMapping("TrainingReplayThread", TrainingReplayThread.class);
```
The new SA TrainingReplayThread class is not needed since it only overrides isHiddenFromExternalView() to return true. You can instead use HiddenJavaThread.class here.

veresov replied:
Done

plummercj commented on src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/memory/FileMapInfo.java:
```diff
@@ -125,8 +126,10 @@ private static void populateMetadataTypeArray(TypeDataBase db) {
     metadataTypeArray[6] = db.lookupType("Method");
-    metadataTypeArray[7] = db.lookupType("ObjArrayKlass");
-    metadataTypeArray[8] = db.lookupType("TypeArrayKlass");
+    metadataTypeArray[9] = db.lookupType("MethodData");
```
It looks like MethodData inheriting from Metadata is not a new change, but has always been the case. I'm surprised this didn't cause any test failures before your changes. Did you end up with test failures after your changes?

veresov replied:
Honestly I don't remember, I think @iklam did these changes.

vnkozlov reviewed (approved):
Looks good.

iklam commented on src/hotspot/share/cds/filemap.cpp:
```diff
@@ -1914,6 +1920,42 @@ bool FileMapHeader::validate() {
                   CompactStrings   ? "enabled" : "disabled");
     return false;
   }
+  if (TrainingData::have_data()) {
+    if (_type_profile_level != TypeProfileLevel) {
+      log_info(cds)("The %s's TypeProfileLevel setting (%d)"
+                    " does not equal the current TypeProfileLevel setting (%d).", file_type,
+                    _type_profile_level, TypeProfileLevel);
+      return false;
+    }
+    if (_type_profile_width != TypeProfileWidth) {
+      log_info(cds)("The %s's TypeProfileWidth setting (%d)"
+                    " does not equal the current TypeProfileWidth setting (%d).", file_type,
+                    (int)_type_profile_width, (int)TypeProfileWidth);
+      return false;
+
+    }
+    if (_bci_profile_width != BciProfileWidth) {
+      log_info(cds)("The %s's BciProfileWidth setting (%d)"
+                    " does not equal the current BciProfileWidth setting (%d).", file_type,
+                    (int)_bci_profile_width, (int)BciProfileWidth);
+      return false;
+    }
+    if (_profile_traps != ProfileTraps) {
+      log_info(cds)("The %s's ProfileTraps setting (%s)"
+                    " does not equal the current ProfileTraps setting (%s).", file_type,
+                    _profile_traps ? "enabled" : "disabled",
+                    ProfileTraps   ? "enabled" : "disabled");
+
+      return false;
+    }
+    if (_spec_trap_limit_extra_entries != SpecTrapLimitExtraEntries) {
+      log_info(cds)("The %s's SpecTrapLimitExtraEntries setting (%d)"
+                    " does not equal the current SpecTrapLimitExtraEntries setting (%d).", file_type,
+                    _spec_trap_limit_extra_entries, SpecTrapLimitExtraEntries);
+      return false;
```
The `log_info(cds)` should be replaced with `MetaspaceShared::report_loading_error`.  (The few `log_info` lines above this block will be fixed in [JDK-8356807](https://bugs.openjdk.org/browse/JDK-8356807))

Also, could you add a new jtreg test case for this? You can see examples in `negativeTests` in the existing AOTFlags.java test case. I think you can add your checks into the new AOTProfileFlags.java test.

veresov replied:
Do you want me to leave the existing `log_info` alone? Or should I fix everything in `FileMapHeader::validate()` ?

iklam replied:
You can leave the existing code and just fix the new code you added.

veresov replied:
Done. And changed the test. Please take a look.

iklam reviewed (approved):
LGTM. Small nits about logging and testing.

DanHeidinga commented on src/hotspot/share/cds/filemap.hpp:
```diff
@@ -117,7 +117,7 @@ class FileMapHeader: private CDSFileMapHeaderBase {
   bool    _compressed_class_ptrs;                 // save the flag UseCompressedClassPointers
   int     _narrow_klass_pointer_bits;             // save number of bits in narrowKlass
-  int     _narrow_klass_shift;                    // save shift width used to pre-compute narrowKlass IDs in archived heap objects
+  int     _narrow_klass_shift;                    // save shift width used to pre-compute narrowKlass IDs in archived heap objectsa
```
```suggestion
  int     _narrow_klass_shift;                    // save shift width used to pre-compute narrowKlass IDs in archived heap objects
```

Minor typo, don't bother fixing if it will result in a re-review cycle

iklam commented on test/hotspot/jtreg/runtime/cds/appcds/aotProfile/AOTProfileFlags.java:
```diff
@@ -0,0 +1,140 @@
+            trainingFlags,
+            "-cp", appJar, helloClass);
+
+        out = CDSTestUtils.executeAndLog(pb, "production_success");
+        out.shouldNotContain("does not equal");
+        out.shouldHaveExitValue(0);
+
+        pb = ProcessTools.createLimitedTestJavaProcessBuilder(
+            "-XX:AOTCache=" + aotCacheFile,
+            "-XX:+UnlockExperimentalVMOptions",
+            productionFlags,
+            "-cp", appJar, helloClass);
+
+        out = CDSTestUtils.executeAndLog(pb, "production_failure");
+        out.shouldContain("does not equal");
```
Since all the flags have `Profile` in them, I think we should use this to match the intended output:

```
String errorPattern = "Profile.* setting .* does not equal the current .*Profile.* setting";
out.shouldNotMatch(errorPattern);
...
out.shouldMatch(errorPattern);
```

veresov replied:
`SpecTrapLimitExtraEntries` does not.

veresov replied:
Fixed. Take a look.

iklam reviewed (approved):
updates are good!

vnkozlov reviewed (approved):
Re-approved

veresov commented:
/integrate
